File: src/main/java/org/jcodec/containers/imgseq/ImageSequenceDemuxer.java
package org.jcodec.containers.imgseq;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.jcodec.common.Codec;
import org.jcodec.common.Demuxer;
import org.jcodec.common.DemuxerTrack;
import org.jcodec.common.DemuxerTrackMeta;
import org.jcodec.common.TrackType;
import org.jcodec.common.io.NIOUtils;
import org.jcodec.common.logging.Logger;
import org.jcodec.common.model.Packet;
import org.jcodec.common.model.Packet.FrameType;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 * 
 * A demuxer that reads image files out of a folder.
 * 
 * Supports both sequences starting with 0 and 1 index.
 * 
 * @author Stanislav Vitvitskyy
 * 
 */
public class ImageSequenceDemuxer implements Demuxer, DemuxerTrack {

    private static final int VIDEO_FPS = 25;
    private String namePattern;
    private int frameNo;
    private Packet curFrame;
    private Codec codec;
    private int maxAvailableFrame;
    private int maxFrames;

    public ImageSequenceDemuxer(String namePattern, int maxFrames) throws IOException {
        this.namePattern = namePattern;
        this.maxFrames = maxFrames;
        this.maxAvailableFrame = -1;
        this.curFrame = loadFrame();
        // codec = JCodecUtil.detectDecoder(curFrame.getData());
        String lowerCase = namePattern.toLowerCase();
        if (lowerCase.endsWith(".png")) {
            codec = Codec.PNG;
        } else if (lowerCase.endsWith(".jpg") || lowerCase.endsWith(".jpeg")) {
            codec = Codec.JPEG;
        }
        // Picture8Bit frame = JCodecUtil.createVideoDecoder(codec,
        // null).decodeFrame8Bit(curFrame.getData(),
        // new byte[3][1920 * 1088]);
        // dimensions = new Size(frame.getWidth(), frame.getHeight());
    }

    @Override
    public void close() throws IOException {
    }

    @Override
    public List<? extends DemuxerTrack> getTracks() {
        ArrayList<DemuxerTrack> tracks = new ArrayList<DemuxerTrack>();
        tracks.add(this);
        return tracks;
    }

    @Override
    public List<? extends DemuxerTrack> getVideoTracks() {
        return getTracks();
    }

    @Override
    public List<? extends DemuxerTrack> getAudioTracks() {
        return new ArrayList<DemuxerTrack>();
    }

    @Override
    public Packet nextFrame() throws IOException {
        try {
            return curFrame;
        } finally {
            curFrame = loadFrame();
        }
    }

    private Packet loadFrame() throws IOException {
        if (frameNo > maxFrames) {
            return null;
        }

        File file = null;
        do {
            String name = String.format(namePattern, frameNo);
            file = new File(name);
            if (file.exists() || frameNo > 0)
                break;
            frameNo++;
        } while (frameNo < 2);

        if (!file.exists())
            return null;

        Packet ret = new Packet(NIOUtils.fetchFromFile(file), frameNo, VIDEO_FPS, 1, frameNo, FrameType.KEY, null, frameNo);
        ++frameNo;
        return ret;
    }

    private static final int MAX_MAX = 60 * 60 * 60 * 24; // Longest possible
                                                          // movie

    /**
     * Finds maximum frame of a sequence by bisecting the range.
     * 
     * Performs at max at max 48 Stat calls ( 2*log2(MAX_MAX) ).
     * 
     * @return
     */
    public int getMaxAvailableFrame() {
        if (maxAvailableFrame == -1) {

            int firstPoint = 0;
            for (int i = MAX_MAX; i >= 0; i /= 2) {
                if (new File(String.format(namePattern, i)).exists()) {
                    firstPoint = i;
                    break;
                }
            }
            int pos = firstPoint;
            for (int interv = firstPoint / 2; interv > 1; interv /= 2) {
                if (new File(String.format(namePattern, pos + interv)).exists()) {
                    pos += interv;
                }
            }
            maxAvailableFrame = pos;
            Logger.info("Max frame found: " + maxAvailableFrame);
        }
        return Math.min(maxAvailableFrame, maxFrames);
    }

    @Override
    public DemuxerTrackMeta getMeta() {
        int durationFrames = getMaxAvailableFrame();
        return new DemuxerTrackMeta(TrackType.VIDEO, codec, (durationFrames + 1) * VIDEO_FPS, null, durationFrames + 1,
                null, null, null);
    }
}

File: src/main/java/org/jcodec/containers/imgseq/ImageSequenceDemuxer.kt
package org.jcodec.containers.imgseq

import org.jcodec.common.*
import org.jcodec.common.io.NIOUtils
import org.jcodec.common.logging.Logger
import org.jcodec.common.model.Packet
import org.jcodec.common.model.Packet.FrameType
import java.io.File
import java.io.IOException
import java.util.*

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed
 * under FreeBSD License
 *
 * A demuxer that reads image files out of a folder.
 *
 * Supports both sequences starting with 0 and 1 index.
 *
 * @author Stanislav Vitvitskyy
 */
class ImageSequenceDemuxer(private val namePattern: String, private val maxFrames: Int) : Demuxer, DemuxerTrack {
    private var frameNo = 0
    private var curFrame: Packet? = null
    private var codec: Codec? = null
    private var _maxAvailableFrame: Int = -1
    private var prevName: String? = null

    @Throws(IOException::class)
    override fun close() {
    }

    override fun getTracks(): List<DemuxerTrack> {
        val tracks = ArrayList<DemuxerTrack>()
        tracks.add(this)
        return tracks
    }

    override fun getVideoTracks(): List<DemuxerTrack> {
        return tracks
    }

    override fun getAudioTracks(): List<DemuxerTrack> {
        return ArrayList()
    }

    @Throws(IOException::class)
    override fun nextFrame(): Packet? {
        return try {
            curFrame
        } finally {
            curFrame = loadFrame()
        }
    }

    @Throws(IOException::class)
    private fun loadFrame(): Packet? {
        if (frameNo > maxFrames) {
            return null
        }
        var file: File? = null
        do {
            val name = String.format(namePattern, frameNo)
            // In case the name doesn't contain placeholders, to prevent infinitely
            // looping around the same file.
            if (name == prevName) {
                return null
            }
            prevName = name
            file = File(name)
            if (file.exists() || frameNo > 0) break
            frameNo++
        } while (frameNo < 2)
        if (file == null || !file.exists()) return null
        val ret = Packet(NIOUtils.fetchFromFile(file), frameNo.toLong(), VIDEO_FPS, 1, frameNo.toLong(), FrameType.KEY, null, frameNo)
        ++frameNo
        return ret
    }
    // movie
    /**
     * Finds maximum frame of a sequence by bisecting the range.
     *
     * Performs at max at max 48 Stat calls ( 2*log2(MAX_MAX) ).
     *
     * @return
     */
    fun getMaxAvailableFrame(): Int {
        if (_maxAvailableFrame == -1) {
            var firstPoint = 0
            var i = MAX_MAX
            while (i > 0) {
                if (File(String.format(namePattern, i)).exists()) {
                    firstPoint = i
                    break
                }
                i /= 2
            }
            var pos = firstPoint
            var interv = firstPoint / 2
            while (interv > 1) {
                if (File(String.format(namePattern, pos + interv)).exists()) {
                    pos += interv
                }
                interv /= 2
            }
            _maxAvailableFrame = pos
            Logger.info("Max frame found: $_maxAvailableFrame")
        }
        return Math.min(_maxAvailableFrame, maxFrames)
    }

    override fun getMeta(): DemuxerTrackMeta {
        val durationFrames = getMaxAvailableFrame()
        return DemuxerTrackMeta(TrackType.VIDEO, codec, ((durationFrames + 1) * VIDEO_FPS).toDouble(), null, durationFrames + 1,
                null, null, null)
    }

    companion object {
        private const val VIDEO_FPS = 25
        private const val MAX_MAX = 60 * 60 * 60 * 24 // Longest possible
    }

    init {
        curFrame = loadFrame()!!
        // codec = JCodecUtil.detectDecoder(curFrame.getData());
        val lowerCase = namePattern.toLowerCase()
        if (lowerCase.endsWith(".png")) {
            codec = Codec.PNG
        } else if (lowerCase.endsWith(".jpg") || lowerCase.endsWith(".jpeg")) {
            codec = Codec.JPEG
        }
    }
}